MismatchGeneration/1_mismatch_generations/Open Library Mismatches/OL-Publication-Date-Mismatch-Generation.py

```diff
@@ -0,0 +1,249 @@
+ 
+    Returns:
+    first_publication_date: This is the first date of publication according to Open Library.
+    """
+    url = f"https://openlibrary.org/works/{external_id}.json"
+    request = requests.get(url)
+    book_data = request.json()
+    
+    if "first_publish_date" in book_data:
+        first_publication_date = book_data["first_publish_date"]
+        return first_publication_date
+    return None
+
+
+# In[5]:
+
+
+# Testing get_first_publication_date().
+print(get_first_publication_date("OL27448W"))
+
+
+# In[6]:
+
+
+def convert_to_datetime(date_str):
+    """
+    This function accepts a publication date and converts it to ISO 8601 format
+    (the format used by Wikidata).
+ 
+    Parameters:
+    date_str (date_str): This is a publication date from the external source.
+ 
+    Returns:
+    iso_date_str: This is the original date_str converted to ISO 8601 format.
+    """
+    if len(date_str) == 4:
+        date_obj = datetime.strptime(date_str, "%Y")
+    elif ',' in date_str:
+        date_obj = datetime.strptime(date_str, "%B %d, %Y")
+    else:
+        date_obj = datetime.strptime(date_str, "%B %Y")
+    # Convert the parsed date object to a string in ISO 8601 format.
+    iso_date_str = date_obj.strftime("%Y-%m-%dT%H:%M:%SZ")
+    return iso_date_str
+
+
+# In[7]:
+
+
+# Testing convert_to_datetime().
+date_str = "1996"
+print(f"1996 is {convert_to_datetime(date_str)} in datetime")
+date_str = "January 1996"
+print(f"January 1996 is {convert_to_datetime(date_str)} in datetime")
+date_str = "January 01, 1996"
+print(f"January 01, 1996 is {convert_to_datetime(date_str)} in datetime")
+date_str = "January 1, 1996"
+print(f"January 1, 1996 is {convert_to_datetime(date_str)} in datetime")
+
+
+# In[8]:
+
+
+mismatches_1 = []
+mismatches_2 = []
+mismatches_3 = []
+
+
+# In[9]:
+
+
+def find_mismatches(wikidata_data, dataframe):
+    """
+    This function accepts a dataframe with the item's Wikidata data and a dataframe which the
+    mismatches will be appended to, and crosschecks the item's publication date on Wikidata
+    with the data on the external source. If the corresponding item has no publication date on
+    the external source, it will not be appended to the dataframe as a mismatch.
+ 
+    Parameters:
+    wikidata_data: The pandas dataframe which contains the Wikidata ID, publication date,
+    and external source ID of the items of interest.
+    dataframe: The pandas dataframe which the mismatches will be appended to.
+    """
+    total_rows = len(wikidata_data)
+    for index, row in tqdm(wikidata_data.iterrows(), total=total_rows, desc="Processing"):
+        external_id = row["openLibraryID"]
+        wd_date = row["publicationDate"]
+        while True:
+            try:
+                pub_date = get_first_publication_date(external_id)
+                if pub_date is None:  # Skip if pub_date is None.
+                    print(f"Publication date not found on Open Library for item {row['item']}")
```

Comment:
It might be helpful if you also had a block of code where it printed "Publication Date Found" or something for the entries that do have a publication date.  When looking at the mismatch generation process, there are a lot of entries that do not have publication dates and it can get a bit confusing at first.